sklearn/utils/tests/test_stats.py

```diff
@@ -137,6 +149,74 @@ def test_weighted_percentile_2d():
     assert_allclose(w_median, p_axis_0)
 
 
+@pytest.mark.parametrize(
+    "array_namespace, device, dtype_name", yield_namespace_device_dtype_combinations()
+)
+@pytest.mark.parametrize(
+    "data, weights, percentile",
+    [
+        # Random 1D array, constant weights
+        (lambda rng: rng.rand(50), np.ones(50), 50),
+        # Random 1D array and random 1D weights
+        (lambda rng: rng.rand(50), lambda rng: rng.rand(50), 75),
+        # Random 2D array and random 2D weights
+        (lambda rng: rng.rand(20, 3), lambda rng: rng.rand(20, 3), 25),
+        # zero-weights and `rank_percentile=0` (#20528)
+        (np.array([0, 1, 2, 3, 4, 5]), np.array([0, 0, 1, 1, 1, 0]), 0),
+        # np.nan's in data and some zero-weights
+        (np.array([np.nan, np.nan, 0, 3, 4, 5]), np.array([0, 1, 1, 1, 1, 0]), 0),
```

Comment:
Please add cases such that `array` has `float32` dtype and `weights` has an integer dtype and check that the result has `float32` dtype.

Reply:
Done